<?xml version="1.0"?>

<robot name="akros" xmlns:xacro="http://ros.org/wiki/xacro">

<!-- **** LINKS **** -->

<!-- NAVIGATION MODULE (BASE_LINK) -->
<link name="base_link">
  <visual>
    <origin xyz="0 0 0" rpy="0 0 0" />
    <geometry>
      <mesh filename="package://mecanum_car/meshes/navigation_module_centered.stl" />
    </geometry>
    <material name="">
      <color rgba="1 1 1 1" />
    </material>
  </visual>
  <collision>
    <origin xyz="0 0 0" rpy="0 0 0" />
    <geometry>
      <box size='0.21 0.15 0.02'/>
    </geometry>
  </collision>
</link>

<!-- BASE_LINK_INERTIAL (WORKAROUND FOR KDL WHICH DOES NOT ACCEPT BASE_LINK WITH INERTIA) -->
<link name="base_link_inertial">
  <inertial>
    <origin xyz="0 0 0" rpy="0 0 0" />
    <mass value="0.001" />
    <inertia ixx="0.01" ixy="0.01" ixz="0.01" iyy="0.01" iyz="0.01" izz="0.01" />
  </inertial>
</link>

<!-- LD06 LASER_LINK -->
<link name="laser_link">
  <inertial>
    <origin xyz="0 0 0" rpy="0 0 0" />
    <mass value="0.001" />
    <inertia ixx="0.01" ixy="0.01" ixz="0.01" iyy="0.01" iyz="0.01" izz="0.01" />
  </inertial>
  <visual>
    <origin xyz="0 0 0" rpy="0 0 0" />
    <geometry>
      <mesh filename="package://mecanum_car/meshes/ld06.stl" />
    </geometry>
    <material name="">
      <color rgba="0 0 0 1" />
    </material>
  </visual>
  <collision>
    <origin xyz="0 0 0" rpy="0 0 0" />
    <geometry>
      <mesh filename="package://mecanum_car/meshes/ld06.stl" />
    </geometry>
  </collision>
</link>

<!-- LASER_FRAME -->
<link name="laser_frame"/>

<!-- T265 TRACKING CAMERA (T265_POSE_FRAME AS WORKAROUND TO KEEP TREE CONTINUOUS)-->
<link name="t265_pose_frame">
  <inertial>
    <origin xyz="0 0 0" rpy="0 0 0" />
    <mass value="0.1" />
    <inertia ixx="0.01" ixy="0.01" ixz="0.01" iyy="0.01" iyz="0.01" izz="0.01" />
  </inertial>
  <visual>
    <origin xyz="0 0 0" rpy="0 0 0" />
    <geometry>
      <mesh filename="package://mecanum_car/meshes/t265.stl" />
    </geometry>
    <material name="">
      <color rgba="0 0 0 1" />
    </material>
  </visual>
  <collision>
    <origin xyz="0 0 0" rpy="0 0 0" />
    <geometry>
      <box size='0.01 0.01 0.01'/>
    </geometry>
  </collision>
</link>

<!-- BASE MODULE (BASE_FOOTPRINT) -->
<link name="base_footprint">
  <inertial>
    <mass value="1.0"/>
    <inertia ixx="0.015" iyy="0.0375" izz="0.0375" ixy="0" ixz="0" iyz="0"/>
  </inertial>
  <visual>
    <origin xyz="0 0 0" rpy="0 0 0" />
    <geometry>
      <mesh filename="package://mecanum_car/meshes/base_module_centered.stl" />
    </geometry>
    <material name="">
      <color rgba="1 1 1 1" />
    </material>
  </visual>
  <collision>
    <origin xyz="0 0 0.0278" rpy="0 0 0" />
    <geometry>
      <box size='0.21 0.15 0.02'/>
    </geometry>
  </collision>
</link>

<!-- WHEEL LEFT FRONT -->
<link name="wheel_lf">
  <inertial>
    <mass value="0.1"/>
    <inertia ixx="5.1458e-5" iyy="5.1458e-5" izz="6.125e-5" ixy="0" ixz="0" iyz="0"/>
  </inertial>
  <visual>
    <origin xyz="0 0 0" rpy="0 0 0" />
    <geometry>
      <mesh filename="package://mecanum_car/meshes/wheel_left_front.stl" />
    </geometry>
    <material name="">
      <color rgba="0 0 0 1" />
    </material>
  </visual>
  <collision>
    <origin xyz="0 0 0" rpy="1.57075 0 0" />
    <geometry>
      <cylinder length="0.02" radius="0.04"/>
    </geometry>
  </collision>
</link>

<!-- WHEEL LEFT BACK -->
<link name="wheel_lb">
  <inertial>
    <mass value="0.1"/>
    <inertia ixx="5.1458e-5" iyy="5.1458e-5" izz="6.125e-5" ixy="0" ixz="0" iyz="0"/>
  </inertial>
  <visual>
    <origin xyz="0 0 0" rpy="0 0 0" />
    <geometry>
      <mesh filename="package://mecanum_car/meshes/wheel_left_back.stl" />
    </geometry>
    <material name="">
      <color rgba="0 0 0 1" />
    </material>
  </visual>
  <collision>
    <origin xyz="0 0 0" rpy="1.57075 0 0" />
    <geometry>
      <cylinder length="0.02" radius="0.04"/>
    </geometry>
  </collision>
</link>

<!-- WHEEL RIGHT FRONT -->
<link name="wheel_rf">
  <inertial>
    <mass value="0.1"/>
    <inertia ixx="5.1458e-5" iyy="5.1458e-5" izz="6.125e-5" ixy="0" ixz="0" iyz="0"/>
  </inertial>
  <visual>
    <origin xyz="0 0 0" rpy="0 0 0" />
    <geometry>
      <mesh filename="package://mecanum_car/meshes/wheel_right_front.stl" />
    </geometry>
    <material name="">
      <color rgba="0 0 0 1" />
    </material>
  </visual>
  <collision>
    <origin xyz="0 0 0" rpy="1.57075 0 0" />
    <geometry>
      <cylinder length="0.02" radius="0.04"/>
    </geometry>
  </collision>
</link>

<!-- WHEEL RIGHT BACK --> 
<link name="wheel_rb">
  <inertial>
    <mass value="0.1"/>
    <inertia ixx="5.1458e-5" iyy="5.1458e-5" izz="6.125e-5" ixy="0" ixz="0" iyz="0"/>
  </inertial>
  <visual>
    <origin xyz="0 0 0" rpy="0 0 0" />
    <geometry>
      <mesh filename="package://mecanum_car/meshes/wheel_right_back.stl" />
    </geometry>
    <material name="">
      <color rgba="0 0 0 1" />
    </material>
  </visual>
  <collision>
    <origin xyz="0 0 0" rpy="1.57075 0 0" />
    <geometry>
      <cylinder length="0.02" radius="0.04"/>
    </geometry>
  </collision>
</link>

<!-- **** JOINTS **** -->

<!-- BASE_LINK_INERTIAL TO BASE_LINK -->
<joint name="base_inertial_fixed" type="fixed">
  <parent link="base_link"/>
  <child link="base_link_inertial"/>
  <origin xyz="0 0 0"/>
</joint>

<!-- LASER_LINK TO BASE_LINK -->
<joint name="base_lidar_fixed" type="fixed">
  <parent link="base_link"/>
  <child link="laser_link"/>
  <origin xyz="0 0 0.0415"/>
</joint>

<!-- LASER_FRAME TO LASER_LINK -->
<joint name="laser_frame_fixed" type="fixed">
  <parent link="laser_link"/>
  <child link="laser_frame"/>
  <origin xyz="0 0 0.03"/>
</joint>

<!-- T265_POSE_FRAME TO BASE_LINK -->
<joint name="base_t265_fixed" type="fixed">
  <parent link="base_link" />
  <child link="t265_pose_frame" />
  <origin xyz="0.092 0 0.011" rpy="0 0 0" />
</joint>

<!-- BASE_FOOTPRINT TO BASE_LINK -->
<joint name="base_link_footprint_fixed" type="fixed">
  <parent link="base_link"/>
  <child link="base_footprint"/>
  <origin xyz="0 0 -0.099" rpy="0 0 0"/>
  <axis xyz="0 0 0"/>
</joint>

<!-- WHEEL_LF TO BASE_FOOTPRINT -->
<joint name="joint_lf" type="continuous">
  <origin xyz="0.051 0.0973 0.0178" rpy="0 0 0" />
  <parent link="base_footprint" />
  <child link="wheel_lf" />
  <axis xyz="0 1 0" />
</joint>
<transmission name="wheel_lf_trans">
  <type>transmission_interface/SimpleTransmission</type>
  <joint name="joint_lf">
    <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
  </joint>
  <actuator name="motor_lf">
    <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
    <mechanicalReduction>1</mechanicalReduction>
  </actuator>
</transmission>

<!-- WHEEL_LB TO BASE_FOOTPRINT -->
<joint name="joint_lb" type="continuous">
  <origin xyz="-0.0475 0.09731 0.0178" rpy="0 0 0" />
  <parent link="base_footprint" />
  <child link="wheel_lb" />
  <axis xyz="0 1 0" />
</joint>
<transmission name="wheel_lb_trans">
  <type>transmission_interface/SimpleTransmission</type>
  <joint name="joint_lb">
    <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
  </joint>
  <actuator name="motor_lb">
    <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
    <mechanicalReduction>1</mechanicalReduction>
  </actuator>
</transmission>

<!-- WHEEL_RF TO BASE_FOOTPRINT -->
<joint name="joint_rf" type="continuous">
  <origin xyz="0.051 -0.0973 0.0178" rpy="0 0 0" />
  <parent link="base_footprint" />
  <child link="wheel_rf" />
  <axis xyz="0 1 0" />
</joint>
<transmission name="wheel_rf_trans">
  <type>transmission_interface/SimpleTransmission</type>
  <joint name="joint_rf">
    <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
  </joint>
  <actuator name="motor_rf">
    <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
    <mechanicalReduction>1</mechanicalReduction>
  </actuator>
</transmission>

<!-- WHEEL_RB TO BASE_FOOTPRINT -->
<joint name="joint_rb" type="continuous">
  <origin xyz="-0.0475 -0.0973 0.0178" rpy="0 0 0" />
  <parent link="base_footprint" />
  <child link="wheel_rb" />
  <axis xyz="0 1 0" />
</joint>
<transmission name="wheel_rb_trans">
  <type>transmission_interface/SimpleTransmission</type>
  <joint name="joint_rb">
    <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
  </joint>
  <actuator name="motor_rb">
    <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
    <mechanicalReduction>1</mechanicalReduction>
  </actuator>
</transmission>

<!-- **** GAZEBO **** -->

<gazebo reference="base_link">
  <material>Gazebo/White</material>
</gazebo>

<gazebo reference="laser_frame">
  <material>Gazebo/Black</material>
</gazebo>

<gazebo reference="t265_pose_frame">
  <material>Gazebo/Black</material>
</gazebo>

<gazebo reference="base_footprint">
  <material>Gazebo/White</material>
</gazebo>

<gazebo reference="wheel_lf">
  <material>Gazebo/Black</material>
</gazebo>

<gazebo reference="wheel_lb">
  <material>Gazebo/Black</material>
</gazebo>

<gazebo reference="wheel_rf">
  <material>Gazebo/Black</material>
</gazebo>

<gazebo reference="wheel_rb">
  <material>Gazebo/Black</material>
</gazebo>

<gazebo reference="laser">
  <sensor type="ray" name="head_ld06_sensor">
    <pose>0 0 0.06 0 0 0</pose>
    <visualize>false</visualize>
    <update_rate>40</update_rate>
    <ray>
      <scan>
        <horizontal>
          <samples>720</samples>
          <resolution>1</resolution>
          <min_angle>-3.14159265</min_angle>
          <max_angle>3.14159265</max_angle>
        </horizontal>
      </scan>
      <range>
        <min>0.2</min>
        <max>30.0</max>
        <resolution>0.01</resolution>
      </range>
      <noise>
        <type>gaussian</type>
        <mean>0.0</mean>
        <stddev>0.01</stddev>
      </noise>
    </ray>
    <plugin name="gazebo_ros_head_ld06_controller" filename="libgazebo_ros_laser.so">
      <topicName>scan</topicName>
      <frameName>laser_frame</frameName>
    </plugin>
  </sensor>
</gazebo>

<gazebo>
  <plugin name="omni_steering" filename="libgazebo_ros_planar_move.so">
    <commandTopic>cmd_vel</commandTopic>
    <odometryTopic>odom</odometryTopic>
    <odometryFrame>odom</odometryFrame>
    <odometryRate>100.0</odometryRate>
    <robotBaseFrame>base_link</robotBaseFrame>
    <publishTF>false</publishTF>
  </plugin>
</gazebo>

<!--<gazebo>
  <plugin name="control" filename="libgazebo_ros_control.so"/>
</gazebo>-->

</robot>
